types/node/index.d.ts

```diff
@@ -6112,7 +6112,7 @@ declare module "http2" {
     }
 
     export interface ServerStreamFileResponseOptions {
-        statCheck?: (stats: fs.Stats, headers: IncomingHttpHeaders, statOptions: StatOptions) => void;
+        statCheck?: (stats: fs.Stats, headers: IncomingHttpHeaders, statOptions: StatOptions) => void|false;
```

Comment:
Would it might make more sense to just make `statCheck` return `void|boolean`?

Reply:
I want to make it explicitly `false` because the node code compares with `=== false` rather than `!statCheck(...)`. I.e. returning things like `true` doesn't make any sense there. What do you think?

Reply:
I think it makes sense to return `true` -- from the source code it looks like `statCheck` is essentially being treated as a predicate function. The `=== false` suggests that boolean return type is expected/preferred, ensuring that if a user uses a void return type function instead, it won't be interpreted to mean "never stream this file".

Reply:
In fact, I would suggest changing it to _only_ `boolean`.

Reply:
I disagree. My interpretation of the document and the code is that `statCheck()` is normally used just as a hook, not as a predicate, for the client to be able to send headers like `last-modified'. But in addition to that, it can also cancel the file transfer by returning `false` from that hook. `false` is just a special value that is checked. It could be a totally different value, say 42, in which case the type would be `void|42`.

I don't want to make it *only* `boolean` because it'll force the client code unnecessarily to return `true` most of the cases.

Reply:
Ok, I'm convinced that `void` is a reasonable return type (given that `statCheck` is intended to be used, like you said, as a hook).

I disagree that `false` is a special value being checked. 42 is an arbitrary value, but `false` is not; it specifically means that some boolean predicate is not satisfied (that predicate being "based on the input arguments, this file should be streamed").

Looking from a user's perspective, to enforce that `statCheck` only returns `false` is to ask them to write a function that looks like

```js
function statCheck1(stats, headers, statOptions): void|false {
  if (stats.lastModified < Date.now() - 1000) { // just toy code
    return;
  } else {
    // mutate headers
    return false;
  }
}
```

when it is more idiomatic to write

```js
function statCheck2(stats, headers, statOptions): boolean {
  if (stats.lastModified < Date.now() - 1000) {
    return true;
  } else {
    // mutate headers
    return false;
  }
}
```

Both functions end up with the same result. Again, the `=== false` just prevents `statCheck1`'s falsey `undefined` return value from being interpreted as false.

Reply:
Normally the client doesn't cancel the transfer. See the example code in the node document: https://nodejs.org/dist/latest-v8.x/docs/api/http2.html

```javascript
function statCheck(stat, headers) {
  headers['last-modified'] = stat.mtime.toUTCString();
}
```

It's just a hook for adding the `last-modified` header.

If we make it only `boolean`, the client must return `true` even in that case. I'm ok with `void|boolean` if you feel strongly though I still prefer `void|false` because returning `true` isn't very meaningful.

Reply:
Changed to `void|boolean`.

Reply:
Yep, I agree now that only `boolean` is not suitable.

It's true that in this particular case, `true` doesn't hold much meaning. That being said, I still advocate `void|boolean`, because if a function returns `false` in some cases, it is generally expected to return `true` in other cases. I would consider a function that sometimes returns false, and other times returns nothing, to be bad style -- that is my sole reservation.

Reply:
That makes sense. Already changed to `void|boolean'.